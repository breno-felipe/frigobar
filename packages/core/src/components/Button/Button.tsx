import React from 'react';
import styled, { DefaultTheme } from 'styled-components';

interface ButtonProps {
  children: React.ReactNode;
  /** Change background-color */
  skin?: 'primary' | 'info' | 'success' | 'danger' | 'warning' | 'neutral';
  /** change the width, height and font-size values */
  size?: 'small' | 'medium' | 'large' | undefined;
  /** Change disabled state of the button */
  disabled?: boolean;
  /** 100% width */
  full?: boolean;
  /** add border-radius circle */
  rounded?: boolean;
  /** large button */
  large?: boolean;
}

const sizes = ({ spacings }: DefaultTheme) => ({
  small: {
    font: 0.875,
    icon: 1,
    padding: {
      top: spacings.xxsmall,
      right: spacings.small,
      bottom: spacings.xxsmall,
      left: spacings.small,
    },
  },
  medium: {
    font: 1,
    icon: 1.275,
    padding: {
      top: spacings.xsmall,
      right: spacings.medium,
      bottom: spacings.xsmall,
      left: spacings.medium,
    },
  },
  large: {
    font: 1.275,
    icon: 1.5,
    padding: {
      top: spacings.small,
      right: spacings.large,
      bottom: spacings.small,
      left: spacings.large,
    },
  },
});

const Button = styled.button<ButtonProps>`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  transition: all 200ms ease-in-out;
  text-decoration: none;

  ${({
    theme,
    skin = 'neutral',
    size = 'small',
    disabled,
    full,
    rounded,
    large,
  }) => {
    const { radius, colors } = theme;

    return `
      background-color: ${colors[skin][500]};
      color: ${colors.neutral[50]};
      font-size: ${large ? '1.125' : sizes(theme)[size].font}rem;
      padding: ${
        large
          ? `${sizes(theme)[size].padding.top * 2}px
             ${sizes(theme)[size].padding.right * 2}px
             ${sizes(theme)[size].padding.bottom * 2}px
             ${sizes(theme)[size].padding.left * 2}px
          `
          : `${sizes(theme)[size].padding.top}px
             ${sizes(theme)[size].padding.right}px
             ${sizes(theme)[size].padding.bottom}px
             ${sizes(theme)[size].padding.left}px
          `
      };
  
      ${
        disabled
          ? `
        background-color: ${colors.neutral[100]};
        color: ${colors.neutral[400]};
        cursor: not-allowed;
      `
          : `
          &:hover {
            box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
          }
          `
      }
  
      ${full ? 'width: 100%;' : ''}
      ${rounded ? `border-radius: ${radius[3]}px` : ''}
    `;
  }};
`;

Button.defaultProps = {
  skin: 'neutral',
  disabled: false,
  size: 'small',
  full: undefined,
  rounded: undefined,
  large: undefined,
};

export default Button;
