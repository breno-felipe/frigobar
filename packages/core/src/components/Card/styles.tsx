import React from 'react';
import styled from 'styled-components';

const Avatar = styled.img<{
  rounded?: boolean;
}>(
  ({ rounded, theme: { radius } }) => `
    border-radius: ${rounded ? `${radius[3]}px` : `${radius[1]}px`};
  `,
);

const Content = styled.div(
  ({ theme: { spacings } }) => `
    padding:
      ${spacings.xsmall}px
      ${spacings.medium}px
      ${spacings.xsmall}px
      ${spacings.medium}px;
  `,
);

const alignDictionary = {
  left: 'flex-start',
  right: 'flex-end',
  'space-between': 'space-between',
  'space-around': 'space-around',
};

const Footer = styled.footer<{
  align?: 'left' | 'right' | 'space-between' | 'space-around';
}>(
  ({ align = 'left', theme: { spacings } }) => `
    display: flex;
    padding:
      ${spacings.xsmall}px
      ${spacings.medium}px
      ${spacings.medium}px
      ${spacings.medium}px;

    justify-content: ${alignDictionary[align]};
  `,
);

const Subtitle = styled.h3`
  font-size: 1rem;
  margin: 0;
`;

const Title = styled.h2`
  font-size: 1.5rem;
  margin: 0;
`;

const Wrapper = styled.div(
  ({ theme: { spacings } }) => `
    margin-top: ${spacings.xsmall}px;
    margin-bottom: ${spacings.xsmall}px;
    overflow: hidden;
    padding-bottom: 56.25%;
    position: relative;
  `,
);

const Image = styled.img`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  width: 100%;
`;

export interface MediaProps {
  /** an image url to display the media */
  src: string;
}

const Media = ({ src, ...props }: MediaProps): JSX.Element => (
  <Wrapper {...props}>
    <Image src={src} />
  </Wrapper>
);

Avatar.displayName = 'Card.Avatar';
Content.displayName = 'Card.Content';
Footer.displayName = 'Card.Footer';
Subtitle.displayName = 'Card.Subtitle';
Title.displayName = 'Card.Title';
Media.displayName = 'Card.Media';

export { Avatar, Content, Footer, Subtitle, Title, Media };
