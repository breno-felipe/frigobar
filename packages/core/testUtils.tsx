import React from 'react';
import PropTypes from 'prop-types';
import {
  render,
  fireEvent,
  waitFor,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from './src/theme';

const Providers = ({ children }: { children?: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Providers.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, fireEvent, waitFor };
