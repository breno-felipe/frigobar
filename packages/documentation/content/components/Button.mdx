---
title: Button
menu: components
---

Button component

```jsx react-live=true
<Button>Button</Button>
```
