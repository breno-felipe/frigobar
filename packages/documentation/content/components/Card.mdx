---
title: Card
menu: components
---

Card component

```jsx react-live=true
<Card>
  <Card.Header avatar="https://dummyimage.com/72x72">
    <Card.Title>Title</Card.Title>
    <Card.Subtitle>Subtitle</Card.Subtitle>
  </Card.Header>
  <Card.Media src="https://dummyimage.com/600x600" />
  <Card.Content>Lorem</Card.Content>
  <Card.Footer>
    <Button skin="primary" style={{ marginRight: 10 }}>
      Button
    </Button>
    <Button skin="primary">Cancel</Button>
  </Card.Footer>
</Card>
```
