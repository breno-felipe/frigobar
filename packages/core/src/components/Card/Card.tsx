import React from 'react';
import styled from 'styled-components';
import Header from './sub-components';
import { Avatar, Content, Footer, Subtitle, Title, Media } from './styles';

interface CardProps {
  /** sets max-width of the card wrapper */
  maxWidth: string;
  children: React.ReactNode;
}

const Wrapper = styled.section<CardProps>(
  ({ maxWidth = '400px', theme: { colors, radius } }) => `
    display: inline-block;
    max-width: ${maxWidth};
    width: 100%;

    border-radius: ${radius[1]}px;
    
    background-color: ${colors.white};
    box-shadow: 0px 2px 5px 0px ${colors.neutral[200]};
  `,
);

const Card = (props: CardProps): JSX.Element => <Wrapper {...props} />;

Card.Content = Content;
Card.Footer = Footer;
Card.Header = Header;
Card.Media = Media;
Card.Avatar = Avatar;
Card.Title = Title;
Card.Subtitle = Subtitle;

Card.displayName = 'Card';

export default Card;
