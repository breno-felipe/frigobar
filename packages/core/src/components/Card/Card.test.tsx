import React from 'react';
import { render } from '../../../testUtils';

import Card from './Card';
import Button from '../Button';

describe('<Card />', () => {
  describe('Snapshots', () => {
    const Component = ({ rounded }: { rounded?: boolean }) => (
      <Card maxWidth="400px">
        <Card.Header
          avatarRounded={rounded}
          avatar="https://dummyimage.com/72x72"
        >
          <Card.Title>Title</Card.Title>
          <Card.Subtitle>Subtitle</Card.Subtitle>
        </Card.Header>
        <Card.Media src="https://dummyimage.com/600x600" />
        <Card.Content>Lorem</Card.Content>
        <Card.Footer>
          <Button skin="primary" style={{ marginRight: 10 }}>
            Button
          </Button>
          <Button skin="primary">Cancel</Button>
        </Card.Footer>
      </Card>
    );

    it('should match snapshot', () => {
      const { container } = render(<Component />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with rounded avatar', () => {
      const { container } = render(<Component rounded />);
      expect(container).toMatchSnapshot();
    });
  });
});
