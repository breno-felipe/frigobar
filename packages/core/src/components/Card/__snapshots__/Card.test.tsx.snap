// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Snapshots should match snapshot 1`] = `
.c2 {
  border-radius: 4px;
}

.c8 {
  padding: 8px 16px 8px 16px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px 16px 16px 16px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c5 {
  font-size: 1rem;
  margin: 0;
}

.c4 {
  font-size: 1.5rem;
  margin: 0;
}

.c6 {
  margin-top: 8px;
  margin-bottom: 8px;
  overflow: hidden;
  padding-bottom: 56.25%;
  position: relative;
}

.c7 {
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 100%;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px 16px 8px 16px;
}

.c3 {
  margin-left: 12px;
}

.c0 {
  display: inline-block;
  max-width: 400px;
  width: 100%;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 0px 2px 5px 0px #ccc;
}

.c10 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #008d96;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c10:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <section
    class="c0"
  >
    <header
      class="c1"
    >
      <img
        class="c2"
        height="72"
        src="https://dummyimage.com/72x72"
        width="72"
      />
      <div
        class="c3"
      >
        <h2
          class="c4"
        >
          Title
        </h2>
        <h3
          class="c5"
        >
          Subtitle
        </h3>
      </div>
    </header>
    <div
      class="c6"
    >
      <img
        class="c7"
        src="https://dummyimage.com/600x600"
      />
    </div>
    <div
      class="c8"
    >
      Lorem
    </div>
    <footer
      class="c9"
    >
      <button
        class="c10"
        style="margin-right: 10px;"
      >
        Button
      </button>
      <button
        class="c10"
      >
        Cancel
      </button>
    </footer>
  </section>
</div>
`;

exports[`<Card /> Snapshots should match snapshot with rounded avatar 1`] = `
.c2 {
  border-radius: 9999px;
}

.c8 {
  padding: 8px 16px 8px 16px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px 16px 16px 16px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c5 {
  font-size: 1rem;
  margin: 0;
}

.c4 {
  font-size: 1.5rem;
  margin: 0;
}

.c6 {
  margin-top: 8px;
  margin-bottom: 8px;
  overflow: hidden;
  padding-bottom: 56.25%;
  position: relative;
}

.c7 {
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 100%;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px 16px 8px 16px;
}

.c3 {
  margin-left: 12px;
}

.c0 {
  display: inline-block;
  max-width: 400px;
  width: 100%;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 0px 2px 5px 0px #ccc;
}

.c10 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #008d96;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c10:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

<div>
  <section
    class="c0"
  >
    <header
      class="c1"
    >
      <img
        class="c2"
        height="72"
        src="https://dummyimage.com/72x72"
        width="72"
      />
      <div
        class="c3"
      >
        <h2
          class="c4"
        >
          Title
        </h2>
        <h3
          class="c5"
        >
          Subtitle
        </h3>
      </div>
    </header>
    <div
      class="c6"
    >
      <img
        class="c7"
        src="https://dummyimage.com/600x600"
      />
    </div>
    <div
      class="c8"
    >
      Lorem
    </div>
    <footer
      class="c9"
    >
      <button
        class="c10"
        style="margin-right: 10px;"
      >
        Button
      </button>
      <button
        class="c10"
      >
        Cancel
      </button>
    </footer>
  </section>
</div>
`;
